fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/properties/FieldWithComparisonCountProperty.java

```diff
@@ -0,0 +1,79 @@
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp.properties;
+
+import com.apple.foundationdb.API;
+import com.apple.foundationdb.record.query.expressions.FieldWithComparison;
+import com.apple.foundationdb.record.query.expressions.QueryComponent;
+import com.apple.foundationdb.record.query.plan.temp.ExpressionRef;
+import com.apple.foundationdb.record.query.plan.temp.PlannerExpression;
+import com.apple.foundationdb.record.query.plan.temp.PlannerProperty;
+import com.apple.foundationdb.record.query.plan.temp.expressions.RelationalPlannerExpression;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * A property that counts the number of {@link FieldWithComparison}s that appear in a planner expression tree.
+ * This property is used as the number of "unsatisfied filters" when picking between query plans that scan different
+ * indexes.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class FieldWithComparisonCountProperty implements PlannerProperty<Integer> {
+    @Override
+    public boolean shouldVisit(@Nonnull PlannerExpression expression) {
+        return expression instanceof RelationalPlannerExpression ||
+               expression instanceof QueryComponent;
+    }
+
+    @Override
+    public boolean shouldVisit(@Nonnull ExpressionRef<? extends PlannerExpression> ref) {
+        return true;
+    }
+
+    @Nonnull
+    @Override
+    public Integer evaluateAtExpression(@Nonnull PlannerExpression expression, @Nonnull List<Integer> childResults) {
+        int total = expression instanceof FieldWithComparison ? 1 : 0;
+        for (Integer childCount : childResults) {
+            if (childCount != null) {
+                total += childCount;
+            }
+        }
+        return total;
+    }
+
+    @Nonnull
+    @Override
+    public Integer evaluateAtRef(@Nonnull ExpressionRef<? extends PlannerExpression> ref, @Nonnull List<Integer> memberResults) {
+        int min = Integer.MAX_VALUE;
+        for (int memberResult : memberResults) {
+            if (memberResult < min) {
+                min = memberResult;
+            }
+        }
+        return min;
+    }
+
+    public static int evaluate(ExpressionRef<? extends PlannerExpression> ref) {
+        return ref.acceptPropertyVisitor(new FieldWithComparisonCountProperty());
```

Comment:
Is there a reason this can't be a static singleton?

Reply:
No. I made it into one.